README.md
# Image Switcher

Switch images to target resolution by breakpoints after JavaScript is loaded.

ImgSwitcher allows you load image by breakpoints and by DPI's (retina display).
You can define how many breakpoints you want to detect, how much DPI you want to check (@2x, @3x, etc.) 
and also you can have a different file type for the original image than the image type that is loaded by ImgSwitcher.

## Installation

You can use NPM package: 

    npm i img-switcher --save

or download this repository and add `imgSwticher.ts` or `imgSwitcher.js` from **public** folder
to your project.


## Usage

At first you must get instance of ImgSwitcher:

    const imgSwitcher = ImgSwitcher.getInstance();

After that you can simply initialize Image Switcher: 

    imgSwitcher.initImgSwitcher();

About naming of files/images, see **Naming** described below.

### Optional Usage

If you want configure options (see **Available options** described below), you can do it via `setOptions()` function:

    imgSwitcher.setOptions(options);

If you need reload list of images (due to lazy load, etc.), you can call `getImages()` function on existing instance:

    imgSwitcher.getImages();

and after that or often you will want to use listener for resize width of viewport, so, 
in these cases you can simply call on existing instance `runImgSwitcher()` function:

    imgSwitcher.runImgSwitcher();

#### Example

    const imgSwitcher = ImgSwitcher.getInstance();
    imgSwitcher.initImgSwitcher();
    
    let timer: Timer;
    window.addEventListener('resize', () => {
        if (timer) {
            clearTimeout(timer);
        }
        timer = setTimeout( () => {
            imgSwitcher.runImgSwitcher();
        }, 333);
    }, true);

### Available options

#### Via JS/TS

Object of options can contains:
 - **attributes** = list of data attributes names, see described bellow
 - **breakpoints** = list of breakpoints, which you want to handle
 - **cssClass** = base class selector for ImgSwitcher
 - **cssStyleClass** = class which is removed from image when is loaded new image
 - **debug** = boolean for enable/disable debugging via console.log for some part of code
 - **multiplier** = list of resolution multipliers, which you want to handle, e.g. retina display (@2x, @3x, etc.)
 
Also, you can set name of data attributes on image for specify separately settings for each of images:
 - **data-img-switcher-bp** = list of breakpoints divided by comma (e.g.: "320,640,960")
 - **data-img-switcher-lbp** = last used breakpoint
 - **data-img-switcher-mp** = list of multipliers divided by comma (e.g.: "1,2,3")
 - **data-img-switcher-te** = target extension it is used for example for SVG elements, where you have to define 
                          which type of new image will be loaded (e.g.: "png", "jpg", etc.)
                          
Example of default options:

    {
        attributes: {
            breakpoints: 'data-img-switcher-bp',
            lastBreakpoint: 'data-img-switcher-lbp',
            multipliers: 'data-img-switcher-mp',
            targetExtension: 'data-img-switcher-te'
        },
        breakpoints: [320, 640, 768, 1024, 1280, 1920],
        cssClass: 'js-img-switcher',
        cssStyleClass: 'o-img-switcher',
        debug: false,
        multiplier: [1, 2]
    }

#### Directly on HTML Tag

Available options, as I described higher, are:
 - breakpoints: `data-img-switcher-bp=""`
    - divided by comma, e.g.: `data-img-switcher-bp="320,640,960"`
 - multipliers: `data-img-switcher-mp=""`
    - divided by comma, e.g.: `data-img-switcher-mp="1,2"`
 - target extension: `data-img-switcher-te=""`
    - it is used when you want to have type of first image another like loaded by ImgSwitcher.
      Typically usage is for SVG as first loaded images. So, if you have SVG (inline or as img tag) and you want to load jpg's
      you must use it in this way:
      
      `<img src="/images/nameOfFile.svg" alt="" title="" data-img-switcher-te="jpg">`
      
      After ImgSwitcher is initialized, he will try to load by specified breakpoint for example `nameOfFile-640.jpg` 

### Naming of files/images

Name of file in situation of three breakpoints (320, 960, 1280) and for @1x, @2x multipliers is:
 - **nameOfImage.jpg** (base most downsized image for first load of page)
 - **nameOfImage-320.jpg** (image until 320px breakpoint)
 - **nameOfImage-320@2x.jpg** (image until 320px breakpoint but for retina displays, so with doubled resolution)
 - **nameOfImage-960.jpg** (image from 321px and until 960px breakpoint)
 - **nameOfImage-960@2x.jpg** (image from 321px and until 960px breakpoint but for retina displays, so with doubled resolution)
 - **nameOfImage-1280.jpg** (image from 961px breakpoint)
 - **nameOfImage-1280@2x.jpg** (image from 961px breakpoint but for retina displays, so with doubled resolution)
 
## Known issues

- *Image is not loaded if script can't find image* 
    - yes, if you forgot add for example `nameOfimage-960.jpg` on your FTP or somewhere, script 
    doesn't try to find closest higher available image. It is because this could be potential overkill for 
    many calls to server, trying to find available variant of image. If you need to have for example four global breakpoints 
    on images but you want just two on one of the images, use data attributes to define breakpoints 
    for this image.
    
If you find a **bug**, please, create a issue. 

## Contributing

### Fork 

1. Fork this project.
2. Clone your fork to your local machine

    `git clone https://github.com/your-username/img-switcher.git`

3. Create a branch with short name described bug/feature

    `git checkout -b branch-name`

4. Make your changes.
   - please, make a **manual testing**
   - **update documentation** (README.md, CHANGELOG.md)
   - **update examples** if it's necessary

5. Commit, push and create a pull request

### Development

Simply run npm install

    npm i
    
After that you can change code and in the end, please, run gulp task runner:

    gulp
    
or

    ./node_modules/.bin/gulp
    
### Rules

Please, install and use TSLint for keeping code style!

### Changelog

There are four base categories:
 - Highlights
    - short description
 - Bug Fixes
    - list of bugs (linked to issue if exists and to commit)
 - Features
    - list of features (linked to feature if exists and to commit)
 - Performance Improvements
    - list of features (linked to feature if exists and to commit)

## Contact

Email: info@striky.sk

Web: https://www.striky.sk/

## License

MIT

CHANGELOG.md
# v1.0.1

### Highlights

Corrections of documentation and added CHANGELOG.md  

# v1.0.0

### Highlights

Welcome ImgSwitcher!

gulpfile.ts
import * as gulp from 'gulp';
import * as gulpTs from 'gulp-typescript';
import * as gulpUglify from 'gulp-uglify';

// source and public folder
const
    source = 'src/',
    dest = 'public/';

function finished(cb: any){
    console.log('\t\t  KEEP \n\r \t\t  CALM \n\r \t\t   and \n\r \t\t  HAPPY \n\r \t\t CODING \n\r \t\t   :)');

    cb();
}

// Compile TS to JS file
gulp.task('compileJStoTS', () => {
    const tsProject = gulpTs.createProject('tsconfig.json');

    return gulp.src(source + '**/*.ts')
        .pipe(tsProject())
        .pipe(gulpUglify())
        .pipe(gulp.dest(dest));
});

// Copy TS files to public
gulp.task('copyTS', () => {
    return gulp.src(source + '**/*.ts')
        .pipe(gulp.dest(dest));
});

gulp.task('default', gulp.series( 'compileJStoTS', 'copyTS', finished));

public/imgSwitcher.js
"use strict";var __assign=this&&this.__assign||Object.assign||function(t){for(var e,i=1,n=arguments.length;i<n;i++)for(var s in e=arguments[i])Object.prototype.hasOwnProperty.call(e,s)&&(t[s]=e[s]);return t};exports.__esModule=!0;var ImgSwitcher=function(){function I(){this.options={attributes:{breakpoints:"data-img-switcher-bp",lastBreakpoint:"data-img-switcher-lbp",multipliers:"data-img-switcher-mp",targetExtension:"data-img-switcher-te"},breakpoints:[320,640,768,1024,1280,1920],cssClass:"js-img-switcher",cssStyleClass:"o-img-switcher",debug:!1,multiplier:[1,2]}}return I.getInstance=function(){return I.instance||(I.instance=new I),I.instance},I.isResMultiplied=function(t){var e=96*t;return t-=.75,window.matchMedia&&(window.matchMedia("only screen and (min-resolution: "+e+"dpi), only screen and (min-resolution: 2dppx), only screen and (min-resolution: "+e/2.54+"dpcm)").matches||window.matchMedia("only screen and (-webkit-min-device-pixel-ratio: "+t+"), only screen and (-o-min-device-pixel-ratio: "+4*t+"/4), only screen and (min--moz-device-pixel-ratio: "+t+"), only screen and (min-device-pixel-ratio: "+t+")").matches)||window.devicePixelRatio&&window.devicePixelRatio>=t},I.getClosestValueCeil=function(t,e){for(var i=null,n=null,s=0,r=e;s<r.length;s++){var a=r[s];if(t<=a&&null===i){i=a;break}a<t&&(null===n||n<a)&&(n=a)}return null!==i?i:n},I.prototype.initImgSwitcher=function(){this.getImages(),this.getMultiplier(),this.callDebug(["Images: "+this.images,"Multiplier: "+this.multiplier]),this.runImgSwitcher()},I.prototype.runImgSwitcher=function(){var w=this;if(0<this.images.length)for(var t=function(t){var e=y.images[t],i=parseInt(e.getAttribute(y.options.attributes.lastBreakpoint),10),n=e.getAttribute(y.options.attributes.targetExtension)?e.getAttribute(y.options.attributes.targetExtension):null,s=0,r="",a=y.multiplier,o=y.options.breakpoints,l="";if("IMG"===e.tagName?(s=e.parentElement.clientWidth,r=e.getAttribute("src")):"svg"===e.tagName?(s=e.parentElement.clientWidth,r=e.getAttribute("data-src")):(s=e.clientWidth,r=getComputedStyle(e).backgroundImage),null===r||""===r.trim()||"none"===r)return"continue";e.hasAttribute(y.options.attributes.multipliers)&&(a=y.getMultiplier(e.getAttribute(y.options.attributes.multipliers).split(",").map(function(t){return parseInt(t,10)}))),1<a&&(l+="@"+a+"x"),e.hasAttribute(y.options.attributes.breakpoints)&&(o=e.getAttribute(y.options.attributes.breakpoints).split(",").map(function(t){return parseInt(t,10)}));var u,c,p,g,m,d,h,b=I.getClosestValueCeil(s,o);if(i!==b){var f=r.replace("-"+i+l,"").replace(/.([^.]+).?$/g,"-"+b+l+(n?"."+n:r.match(/.([^.]+).?$/g)[0])),v=new Image;e.setAttribute(y.options.attributes.lastBreakpoint,b.toString()),"IMG"===e.tagName?(v.src=f,v.onload=(d=e,h=f,function(){d.setAttribute("src",h),d.classList.remove(w.options.cssStyleClass),v=null})):"svg"===e.tagName?(v.src=f,v.onload=(g=e,m=f,function(){var t=document.createElement("img");t.src=m;for(var e=0;e<g.classList.length;e++)t.classList.add(g.classList[e]);for(t.classList.remove(w.options.cssStyleClass),e=0;e<g.attributes.length;e++){var i=g.attributes.item(e);"data-title"===i.name||"data-alt"===i.name?t.setAttribute(i.name.replace("data-",""),i.nodeValue):0<=i.name.indexOf("data-")&&t.setAttribute(i.nodeName,i.nodeValue)}g.parentNode.insertBefore(t,g),g.remove(),t=v=null})):(f=f.replace("url(","").replace(")","").replace(/"/g,"").replace(/'/g,""),v.src=f,v.onload=(u=e,c=f,p=y.options,function(){u.style.backgroundImage="url("+c+")",u.classList.remove(p.cssStyleClass),v=null})),f=null}},y=this,e=0;e<this.images.length;e++)t(e);else this.callDebug(["Init Error: No images were found!"])},I.prototype.setOptions=function(t){this.options=__assign({},this.options,t)},I.prototype.getImages=function(){return this.images=document.querySelectorAll("."+this.options.cssClass)},I.prototype.getMultiplier=function(t){void 0===t&&(t=this.options.multiplier);var e=1;if(1===t.length)e=t[0];else{t.sort(function(t,e){return e-t});for(var i=0,n=t;i<n.length;i++){var s=n[i];if(I.isResMultiplied(s)){e=s;break}}}return this.multiplier=e},I.prototype.callDebug=function(t){if(this.options.debug)for(var e=0,i=t;e<i.length;e++){var n=i[e];console.log(n)}},I}();exports.ImgSwitcher=ImgSwitcher;

src/imgSwitcher.ts
import Timer = NodeJS.Timer;

export interface ImgSwitcherOptions {
    attributes: ImgSwitcherAttributes;
    breakpoints?: Array<number>;
    cssClass?: string;
    cssStyleClass?: string;
    debug?: boolean;
    multiplier?: Array<number>;
}

export interface ImgSwitcherAttributes {
    breakpoints: string;
    lastBreakpoint: string;
    multipliers: string;
    targetExtension: string;
}

/**
 * Image Switcher
 *
 * On IMG or SVG tag is automatically loaded closest best resolution of image by width of his parent.
 * On DIV tag is closest best resolution by width of itself.
 *
 * Name of file in situation of three breakpoints (320, 960, 1280) and for @1x, @2x multipliers is:
 *      - nameOfImage.jpg (base most downsized image for first load of page)
 *      - nameOfImage-320.jpg (image until 320px breakpoint)
 *      - nameOfImage-320@2x.jpg (image until 320px breakpoint but for retina displays, so with doubled resolution)
 *      - nameOfImage-960.jpg (image from 321px and until 960px breakpoint)
 *      - nameOfImage-960@2x.jpg (image from 321px and until 960px breakpoint but for retina displays, so with doubled resolution)
 *      - nameOfImage-1280.jpg (image from 961px breakpoint)
 *      - nameOfImage-1280@2x.jpg (image from 961px breakpoint but for retina displays, so with doubled resolution)
 *
 * Possible global options:
 *      - attributes = list of name data attributes, see described bellow
 *      - breakpoints = list of breakpoints, which you want to handle
 *      - cssClass = base class selector for ImgSwitcher
 *      - cssStyleClass = class which is removed from image when is loaded new image
 *      - debug = boolean for enable/disable echo to console.log some part of code for debugging
 *      - multiplier = list of resolution multipliers, which you want to handle, e.g. retina display (@2, @3, etc.)
 *
 * Also, you can set name of data attributes on image for specify separately settings for each of images:
 *      - data-img-switcher-bp = list of breakpoints divided by comma (e.g.: "320,640,960")
 *      - data-img-switcher-lbp = last used breakpoint
 *      - data-img-switcher-mp = list of multipliers divided by commma (e.g.: "1,2,3")
 *      - data-img-switcher-te = target extension it is used for example for SVG elements, where you must to define
 *                               which type of new image will be loaded (e.g.: "png", "jpg", etc.)
 *
 * @author Lukáš Strišovský, 2018 <info@striky.sk>
 * @license The MIT License, see LICENSE for more information
 */
export class ImgSwitcher {
    private static instance: ImgSwitcher;
    private options: ImgSwitcherOptions = {
        attributes: {
            breakpoints: 'data-img-switcher-bp',
            lastBreakpoint: 'data-img-switcher-lbp',
            multipliers: 'data-img-switcher-mp',
            targetExtension: 'data-img-switcher-te'
        },
        breakpoints: [320, 640, 768, 1024, 1280, 1920],
        cssClass: 'js-img-switcher',
        cssStyleClass: 'o-img-switcher',
        debug: false,
        multiplier: [1, 2]
    };
    private images: NodeListOf<HTMLElement>;
    private multiplier: number;

    private constructor() {}

    public static getInstance(): ImgSwitcher {
        if (!ImgSwitcher.instance) {
            ImgSwitcher.instance = new ImgSwitcher();
        }
        return ImgSwitcher.instance;
    }

    public static isResMultiplied(pixelRatio: number): boolean {
        const dpi = pixelRatio * 96;
        pixelRatio = pixelRatio - 0.75;

        return ((window.matchMedia &&
                (window.matchMedia(
                    'only screen and (min-resolution: ' + dpi + 'dpi), only screen and (min-resolution: 2dppx), ' +
                    'only screen and (min-resolution: ' + (dpi / 2.54) + 'dpcm)').matches
                    || window.matchMedia('only screen and (-webkit-min-device-pixel-ratio: ' + pixelRatio + '), ' +
                        'only screen and (-o-min-device-pixel-ratio: ' + (pixelRatio * 4) + '/4), ' +
                        'only screen and (min--moz-device-pixel-ratio: ' + pixelRatio + '), ' +
                        'only screen and (min-device-pixel-ratio: ' + pixelRatio + ')').matches)
            )
            || (window.devicePixelRatio && window.devicePixelRatio >= pixelRatio));
    }

    private static getClosestValueCeil(width: number, breakpoints: Array<number>): number {
        let higherMinValue = null,
            lowerMaxValue = null;

        for (const breakpoint of breakpoints) {
            if (breakpoint >= width && higherMinValue === null) {
                higherMinValue = breakpoint;
                break;
            }

            if (breakpoint < width && (lowerMaxValue === null || breakpoint > lowerMaxValue)) {
                lowerMaxValue = breakpoint;
            }
        }

        return higherMinValue !== null ? higherMinValue : lowerMaxValue;
    }

    public initImgSwitcher() {
        this.getImages();
        this.getMultiplier();

        this.callDebug([
            'Images: ' + this.images,
            'Multiplier: ' + this.multiplier
        ]);

        this.runImgSwitcher();
    }

    public runImgSwitcher() {
        if (this.images.length > 0) {
            for (let i = 0; i < this.images.length; i++) {
                const image = this.images[i],
                    imageLastBreakpoint = parseInt(image.getAttribute(this.options.attributes.lastBreakpoint), 10),
                    imageTargetExtension = image.getAttribute(this.options.attributes.targetExtension) ? image.getAttribute(this.options.attributes.targetExtension) : null;

                let imageWidth = 0,
                    imageSrc = '',
                    imageMultiplier = this.multiplier,
                    imageBreakpoints = this.options.breakpoints,
                    imageMultiplierString = '';

                // Get Element width and src by type of Element
                if (image.tagName === 'IMG') {
                    imageWidth = image.parentElement.clientWidth;
                    imageSrc = image.getAttribute('src');
                } else if (image.tagName === 'svg') {
                    imageWidth = image.parentElement.clientWidth;
                    imageSrc = image.getAttribute('data-src');
                } else {
                    imageWidth = image.clientWidth;
                    imageSrc = getComputedStyle(image).backgroundImage;
                }

                // Check if we have probably correct src, if not, skip this image
                if (imageSrc === null || imageSrc.trim() === '' || imageSrc === 'none') {
                    continue;
                }

                // Try find separately defined multipliers
                if (image.hasAttribute(this.options.attributes.multipliers)) {
                    imageMultiplier = this.getMultiplier(image.getAttribute(this.options.attributes.multipliers).split(',').map((multiplier) => parseInt(multiplier, 10)));
                }

                // Add multiplier to name of file
                if (imageMultiplier > 1) {
                    imageMultiplierString += '@' + imageMultiplier + 'x';
                }

                // Try find separately defined breakpoints
                if (image.hasAttribute(this.options.attributes.breakpoints)) {
                    imageBreakpoints = image.getAttribute(this.options.attributes.breakpoints).split(',').map((multiplier) => parseInt(multiplier, 10));
                }

                // Get target breakpoint/width
                const targetBreakpoint = ImgSwitcher.getClosestValueCeil(imageWidth, imageBreakpoints);

                // If we have another width as we need, try change it
                if (imageLastBreakpoint !== targetBreakpoint) {
                    // Remove old resolution from src and set new one
                    let newImageSrc = imageSrc
                            .replace('-' + imageLastBreakpoint + imageMultiplierString, '')
                            .replace(/.([^.]+).?$/g, '-' + targetBreakpoint + imageMultiplierString +
                                (imageTargetExtension ? '.' + imageTargetExtension : imageSrc.match(/.([^.]+).?$/g)[0])),
                        newImage = new Image();

                    // Save new resolution to attribute
                    image.setAttribute(this.options.attributes.lastBreakpoint, targetBreakpoint.toString());

                    // Create new temporary image and load him, after that switch loaded image with current
                    if (image.tagName === 'IMG') {
                        newImage.src = newImageSrc;
                        newImage.onload = ((img, newImgSrc) => () => {
                                img.setAttribute('src', newImgSrc);
                                img.classList.remove(this.options.cssStyleClass);

                                newImage = null;
                            }
                        )(image, newImageSrc);
                    } else if (image.tagName === 'svg') {
                        newImage.src = newImageSrc;
                        newImage.onload = ((img, newImgSrc) => () => {
                                // In SVG situation we must create new img Element
                                let newChildImg = document.createElement('img');
                                newChildImg.src = newImgSrc;

                                // Preserve all classes
                                for (let y = 0; y < img.classList.length; y++) {
                                    newChildImg.classList.add(img.classList[y]);
                                }

                                newChildImg.classList.remove(this.options.cssStyleClass);

                                // and also attributes
                                for (let y = 0; y < img.attributes.length; y++) {
                                    const attr = img.attributes.item(y);

                                    // expect data-title/alt which are after change as majority title/alt attributes
                                    if (attr.name === 'data-title' || attr.name === 'data-alt') {
                                        newChildImg.setAttribute(attr.name.replace('data-', ''), attr.nodeValue);
                                    } else if (attr.name.indexOf('data-') >= 0) {
                                        newChildImg.setAttribute(attr.nodeName, attr.nodeValue);
                                    }
                                }

                                // Insert new Element before previous image and remove previous one
                                img.parentNode.insertBefore(newChildImg, img);
                                img.remove();

                                newImage = null;
                                newChildImg = null;
                            }
                        )(image, newImageSrc);
                    } else {
                        newImageSrc = newImageSrc.replace('url(', '').replace(')', '').replace(/"/g, '').replace(/'/g, '');
                        newImage.src = newImageSrc;
                        newImage.onload = ((img, newImgSrc, options) => () => {
                                img.style.backgroundImage = 'url(' + newImgSrc + ')';
                                img.classList.remove(options.cssStyleClass);

                                newImage = null;
                            }
                        )(image, newImageSrc, this.options);
                    }

                    newImageSrc = null;
                }
            }
        } else {
            this.callDebug([
                'Init Error: No images were found!'
            ]);
        }
    }

    public setOptions(_options: ImgSwitcherOptions) {
        this.options = {...this.options, ..._options};
    }

    public getImages(): NodeListOf<HTMLElement> {
        return this.images = (document.querySelectorAll('.' + this.options.cssClass) as NodeListOf<HTMLElement>);
    }

    private getMultiplier(multipliers: Array<number> = this.options.multiplier): number {
        let result = 1;

        if (multipliers.length === 1) {
            result = multipliers[0];
        } else {
            multipliers.sort((a, b) => b - a);

            for (const multiplier of multipliers) {
                if (ImgSwitcher.isResMultiplied(multiplier)) {
                    result = multiplier;
                    break;
                }
            }
        }

        return this.multiplier = result;
    }

    private callDebug(messages: Array<any>) {
        if (this.options.debug) {
            for (const message of messages) {
                console.log(message);
            }
        }
    }
}

const imgSwitcher = ImgSwitcher.getInstance();
imgSwitcher.initImgSwitcher();

let timer: Timer;
window.addEventListener('resize', () => {
    if (timer) {
        clearTimeout(timer);
    }
    timer = setTimeout( () => {
        imgSwitcher.runImgSwitcher();
    }, 333);
}, true);
